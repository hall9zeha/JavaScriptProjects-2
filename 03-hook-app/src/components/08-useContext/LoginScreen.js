import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const LoginScreen = () => {
    const {user, setUser} = useContext(UserContext);

    const actionLogin = ()=>{
        const {id,name,email} = user;
        setUser({
            id,
            name,
            email,
            sigIn:true
        })
    }

  return (
    <div>
        <h1>LoginScreen</h1>
        <hr/>
        <button onClick={actionLogin} className='btn btn-primary '>Login</button>
    </div>
  )
}
