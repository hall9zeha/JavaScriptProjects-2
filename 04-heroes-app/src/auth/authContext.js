import { createContext } from "react";

//Como la devolución de createContext() es un componente, el nombre de la constante 
//tendrá la sintaxis de un Componente(con mayúsculas por convención), esto es opcional claro
export const AuthContext = createContext()