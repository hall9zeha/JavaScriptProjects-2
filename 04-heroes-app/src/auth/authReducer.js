import { Types } from "../types/Types";

export const authReducer = (state ={}, action)=>{

    switch (action.type) {
        case Types.login:
                return {
                    ...action.payload,
                    logged:true
                }
        case Types.logout:
            return {
                logged:false
            }
               
        default:
            return state;
    }
}