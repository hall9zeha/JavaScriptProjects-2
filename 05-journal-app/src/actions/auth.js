// Las acciones no son más que simples funciones que representan nuestros casos de uso

import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/Types";
import { uiFinishLoading, uiStartLoading } from "./ui";



export const startLoginWithEmailPassword = (email, password) =>{
    return (dispatch)=>{
        dispatch(uiStartLoading())
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName));
                dispatch(uiFinishLoading())
            })
            .catch(e=>{
                dispatch(uiFinishLoading())
                console.log(e)
            })
        
    }
}

export const startLoginWithGoogle = ()=>{
    return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName))
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password,name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then(async({user}) =>{
                await user.updateProfile({displayName:name});
                dispatch(login(user.uid,user.displayName));
            })
            .catch(e=>{
                console.log(e)
            })
    }
}


// La siguiente función retornará un objeto
export const login = (uid, displayName)=>({
    type: types.login,
    payload:{
        uid,
        displayName
    }
})

// Logout

export const startLogout = ()=>{
    return async (dispatch)=>{
        await firebase.auth().signOut();
        dispatch(logout())
    }
}

export const logout=()=>({
    type:types.logout
})