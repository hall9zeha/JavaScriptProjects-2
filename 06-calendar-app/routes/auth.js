/*
User routes /auth
host + /api/auth

*/

const {Router} = require('express');
const router = Router();
const {check} = require('express-validator')

const {createUser, login,renewToken} = require('../controllers/auth');
const { fieldValidate } = require('../middlewares/field-validators');
const { validateJWT } = require('../middlewares/jws-validator');

router.post('/new',
    //Midlewares para validar los campos necesarios, comenzamos a hacerlos aqu√≠ y los recibimos en nuestros controllers

    [
        check('name', 'Name is required').not().isEmpty(),
        check('email', 'Email is required').isEmail(),
        check('password', 'The password must have 6 or more characters').isLength({min:6}),
        fieldValidate
    ],    
    createUser)
router.post('/',
    [
        check('email', 'Email is required').isEmail(),
        check('password', 'The password must have 6 or more characters').isLength({min:6}),
        fieldValidate
    ],
    login)
    // Validamos el json web token con vlidateJWT 
router.get('/renew',validateJWT, renewToken)

module.exports = router;