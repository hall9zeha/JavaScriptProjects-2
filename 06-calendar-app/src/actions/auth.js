import Swal from "sweetalert2";
import { fetchWithToken, fetchWithoutToken } from "../helpers/fetch"
import { types } from "../types/types";

export const startLogin = (email, password) =>{
    //Gracias a la librería thunk podemos disponer de un dispatch
    return async(dispatch) =>{
        const resp = await fetchWithoutToken('auth',{email,password},'POST');
        const body = await resp.json();
        
        //Si recibimos un arespuesta correcta de nuestra api 
        /*
            {
                ok:true
                ...
            }
        guardamos el token y la fecha de creación en local storage
        */
       
        if(body.ok){
            localStorage.setItem('token',body.token);
            localStorage.setItem('token-init-date',new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))
        }else{
            Swal.fire('Error',body.msg, 'error');
        }
    }
}

export const startRegister = (name,email,password) =>{
    return async(dispatch)=>{
        const resp = await fetchWithoutToken('auth/new',{name,email,password},'POST');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))

        }else{
            Swal.fire('Error',body.msg, 'error');
        }

    }
}
export const startChecking =()=> {
    return async(dispatch)=>{

        //No requiere ningún argumento porque por defecto será una petición GET
        const resp = await fetchWithToken('auth/renew');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))

        }else{
            Swal.fire('Error',body.msg, 'error');
            dispatch(checkingFinish());
        }
    }
}

const checkingFinish = () =>({type:types.authCheckingFinish});
//Ya que usaremos esta acción solo en este lugar no la exportaremos
const login = (user) =>({
    type:types.authLogin,
    payload:user
})

