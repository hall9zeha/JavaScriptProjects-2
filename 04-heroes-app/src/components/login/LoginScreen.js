import React, { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../../auth/authContext';
import { Types } from '../../types/Types';

export const LoginScreen = () => {
    const navigate = useNavigate();

    const {dispatch} = useContext(AuthContext);

    const handleLogin=()=>{
      const action = {
        type:Types.login,
        payload:{name:'Martha'}
      }
      dispatch(action);
      navigate('/marvel',{
          replace:true
      })
    }

  return (
    <div className='container mt-5'>
        <h1>Login</h1>
        <hr/>

        <button className='btn btn-primary ' onClick={handleLogin}>Login</button>
    </div>
  )
}
