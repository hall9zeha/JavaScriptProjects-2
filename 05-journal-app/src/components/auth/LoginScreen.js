import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { useDispatch } from 'react-redux'
import { login, startLoginAsyncExample, startLoginWithGoogle } from '../../actions/auth'

export const LoginScreen = () => {

  const dispatch = useDispatch()


  const [formValues, handleInputChange ] = useForm({
    email:'martha@gmail.com',
    password:12345678
  })

  const {email,password} = formValues

  const handleEventLogin = (e) =>{
    e.preventDefault();
    dispatch(login(123456,'Martha'))
    
    // Ejemplo de acción asíncrona para probar que la configuración middleware en redux funciona correctamente
    // dispatch(startLoginAsyncExample(email,password));
  }
  const handleLoginWithGoogle =()=>{
    dispatch(startLoginWithGoogle());
  }
  return (
    <>
      <h3 className='auth__title'>Login</h3>
      <form onSubmit={handleEventLogin}>
        <input
          type='text'
          placeholder='Email'
          name="email"
          autoComplete='off'
          className='auth__input'
          value={email}
          onChange={handleInputChange}
        />
        <input
          type='password'
          placeholder='Password'
          name='password'
          className='auth__input'
          value={password}
          onChange={handleInputChange}
        />
        <button

          type='submit'
          className='btn btn-primary btn-block'
        >
          Login
        </button>
    
        <div className='auth__social-networks'>
          <p>Login with social networks</p>
              <div  className="google-btn" 
                    onClick={handleLoginWithGoogle}
              >
                  <div className="google-icon-wrapper">
                      <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/archive/c/c1/20210618182605%21Google_%22G%22_logo.svg" alt="google button" />
                  </div>
                  <p className="btn-text">
                      <b>Sign in with google</b>
                  </p>
              </div>
        </div>
        <Link
          to="/auth/register"
          className="link"
        >
          Create new account
        </Link>   
      </form>
    </>
  )
}
