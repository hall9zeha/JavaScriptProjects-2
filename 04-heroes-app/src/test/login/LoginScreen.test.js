import renderer from 'react-test-renderer';
import { AuthContext } from '../../auth/authContext';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { LoginScreen } from '../../components/login/LoginScreen';

describe('Tests in <LoginScreen/> component', () => { 

    const contextValue = {
        dispatch:jest.fn(),
        user:{
            name:'Barry',
            logged:false
        }
    }
    test('should show successfully', () => { 
        const wrapper = renderer.create(
            <AuthContext.Provider value={contextValue}>
                 <MemoryRouter initialEntries={['/']}>
                    <Routes>
                        <Route path='/' element={<LoginScreen/>}/>
                    </Routes>
                </MemoryRouter>
            </AuthContext.Provider>
        )
        expect(wrapper).toMatchSnapshot();
     })
 })