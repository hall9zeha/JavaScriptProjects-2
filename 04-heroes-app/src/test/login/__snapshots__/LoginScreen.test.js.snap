// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <LoginScreen/> component should show successfully 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary "
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
